eras/babbage/impl/src/Cardano/Ledger/Babbage/PParams.hs

```diff
@@ -0,0 +1,533 @@
+        <! From -- _maxBHSize       :: Natural
+        <! From -- _keyDeposit      :: Coin
+        <! From -- _poolDeposit     :: Coin
+        <! From -- _eMax            :: EpochNo
+        <! From -- _nOpt            :: Natural
+        <! From -- _a0              :: NonNegativeInterval
+        <! From -- _rho             :: UnitInterval
+        <! From -- _tau             :: UnitInterval
+        <! D fromCBORGroup -- _protocolVersion :: ProtVer
+        <! From -- _minPoolCost     :: Natural
+        <! From -- _coinsPerUTxOWord  :: Coin
+        <! D decodeCostModelMap -- _costmdls :: (Map Language CostModel)
+        <! From -- _prices = prices',
+        <! From -- _maxTxExUnits = maxTxExUnits',
+        <! From -- _maxBlockExUnits = maxBlockExUnits'
+        <! From -- maxValSize :: Natural
+        <! From -- collateralPercentage :: Natural
+        <! From -- maxCollateralInputs :: Natural
+
+-- | Returns a basic "empty" `PParams` structure with all zero values.
+emptyPParams :: PParams era
+emptyPParams =
+  PParams
+    { _minfeeA = 0,
+      _minfeeB = 0,
+      _maxBBSize = 0,
+      _maxTxSize = 2048,
+      _maxBHSize = 0,
+      _keyDeposit = Coin 0,
+      _poolDeposit = Coin 0,
+      _eMax = EpochNo 0,
+      _nOpt = 100,
+      _a0 = minBound,
+      _rho = minBound,
+      _tau = minBound,
+      _protocolVersion = BT.ProtVer 0 0,
+      _minPoolCost = mempty,
+      _coinsPerUTxOWord = Coin 0,
+      _costmdls = mempty,
+      _prices = Prices minBound minBound,
+      _maxTxExUnits = ExUnits 0 0,
+      _maxBlockExUnits = ExUnits 0 0,
+      _maxValSize = 0,
+      _collateralPercentage = 150,
+      _maxCollateralInputs = 5
+    }
+
+-- | Since ExUnits does not have an Ord instance, we have to roll this Ord instance by hand.
+-- IF THE ORDER OR TYPES OF THE FIELDS OF PParams changes, this instance may need adusting.
+instance Ord (PParams' StrictMaybe era) where
+  compare x y =
+    (_minfeeA x, _minfeeA y)
+      <== (_minfeeB x, _minfeeB y)
+      <== (_maxBBSize x, _maxBBSize y)
+      <== (_maxTxSize x, _maxTxSize y)
+      <== (_maxBHSize x, _maxBHSize y)
+      <== (_keyDeposit x, _keyDeposit y)
+      <== (_poolDeposit x, _poolDeposit y)
+      <== (_eMax x, _eMax y)
+      <== (_nOpt x, _nOpt y)
+      <== (_a0 x, _a0 y)
+      <== (_rho x, _rho y)
+      <== (_tau x, _tau y)
+      <== (_protocolVersion x, _protocolVersion y)
+      <== (_minPoolCost x, _minPoolCost y)
+      <== (_coinsPerUTxOWord x, _coinsPerUTxOWord y)
+      <== (_costmdls x, _costmdls y)
+      <== (_prices x, _prices y)
+      <== ( case compareEx (_maxTxExUnits x) (_maxTxExUnits y) of
+              LT -> LT
+              GT -> GT
+              EQ -> case compareEx (_maxBlockExUnits x) (_maxBlockExUnits y) of
+                LT -> LT
+                GT -> GT
+                EQ -> (_maxValSize x, _maxValSize y) <== EQ
+          )
+
+infixr 4 <==
+
+(<==) :: Ord a => (a, a) -> Ordering -> Ordering
```

Comment:
I wonder if we can find a good home for this function, Alonzo uses it as well.

Reply:
We don'y really need this function, this is equivalent to `<>` on `Ordering`.